### PR TITLE
lvm, sub uid/gid, lxc 2.1

### PR DESCRIPTION
**Pull request by slmi**

### CODE REVIEW

**Comment by globalcitizen** [comment]
Looks like good work. Have you tested it with earlier versions and without using LVM and validated it works? If so I'll take your word for it and merge. I don't have a test configuration at present and am about to enter a travel vortex.

File: lxc-gentoo [hunk 1]
```diff
@@ -952,6 +958,44 @@ create()
...
+	    }
+	    if ($2 < subgid ) {
+		print "chgrp -h "subgid + $2" -- '\''"$3"'\'';"
+	    }
+	    if ( $5 != "symbolic link" ) {
+		print "chmod -fc "$4" -- '\''"$3"'\'';"
+	    }
+	  }' | /bin/sh
+
+	# Permition for sub uid/gid
```
Comment (specing) [inline lxc-gentoo]: typo

File: lxc-gentoo [hunk 2]
```diff
@@ -952,6 +958,44 @@ create()
 	# never hurts to have a fail-safe.
 	[[ -n "${NAME//\/}" ]] \
 		|| die 8 "\$NAME (%s) IS EMPTY OR MADE OF ONLY DIRECTORY SEPERATORS, THIS IS *VERY* BAD!\n" "$NAME"
+apply_subid()
+{
+    printf "Apply sub_uid/sub_gid\n"
+    if [ -n "$SUB_UID" ] && [ -n "$SUB_GID" ];then
+	cd $ROOTFS
```
Comment (specing) [inline lxc-gentoo]: What if ROOTFS contains spaces?

File: lxc-gentoo [hunk 3]
```diff
@@ -951,9 +957,31 @@ configure()
...
+	    FSSIZE="5G"
+	fi
+	lvcreate -n $LVNAME -L $FSSIZE $VGNAME \
+	    || die 2 "Error: create lvm\n"
+	mkfs.$FSTYPE -L $NAME /dev/$VGNAME/$LVNAME \
+	    || die 2 "Error: format lvm\n"
+	mount /dev/$VGNAME/$LVNAME $ROOTFS \
+	    || die 2 "Error: mount lvm\n"
+    else
+	die 1 "Error: BDEV invlaide lvm"
```
Comment (specing) [inline lxc-gentoo]: typo

File: lxc-gentoo [hunk 4]
```diff
@@ -1132,33 +1140,33 @@ destroy()
...
+			 Current/Default: ${GUESTROOTPASS}
+  -m MIRROR		mirror for stage3 and Portage archive
+			 Current/Default: ${MIRROR}
+  -t STAGE3_TARBALL	path to a stage3 archive to use
+			 Specifying this option will avoid fetching stage (or asking for arch, variant and mirror)
+			 Current: ${STAGE3_TARBALL} Default: null
+  -P PORTAGE_SOURCE	path to a custom portage to use. Can be one of:
 			 - tarball -- will be extracted
-			 - directory -- will be bind-mounted read-only
+			 - directory -- will be bind-mounted read-only (/usr/portage/tree)
```
Comment (specing) [inline lxc-gentoo]: add 'e.g.' if an example must be added.

**Comment by slmi** [comment]
No functional with lxc <2.1 ("Network configuration"  and "Table of changed configuration keys"-> https://discuss.linuxcontainers.org/t/lxc-2-1-has-been-released/487/1)
example: 
> lxc.network. -> lxc.net.[i]. 
> lxc.network.ipv4 -> lxc.net.[i].ipv4.address

Check, fix and result:
# nolvm(-B) sub uid/gid(-s):
```
# ./lxc-gentoo create -n test-nolvm -u test-nolvm.local -a amd64 -q -s 10000000:10000000 -i 192.168.3.99/24 -g 192.168.3.254 -P /usr/portage/tree -r ./test/rootfs -c ./test/config
NAME           = test-nolvm
UTSNAME        = test-nolvm.local
ROOTFS         = ./test/rootfs
CONFFILE       = ./test/config
ARCH           = amd64
ARCHVARIANT    = amd64
STAGE3_TARBALL = 
GUESTROOTPASS  = (not shown)
MIRROR         = http://distfiles.gentoo.org
PORTAGE_SOURCE = /usr/portage/tree
CACHE          = /var/cache/lxc/gentoo
IPV4           = 192.168.3.99/24
IPV4_GATEWAY   = 192.168.3.254
IPV6           = dhcp
IPV6_GATEWAY   = auto
Thanks! Now sit back and relax while your gentoo brews...


Attempting to obtain an exclusive lock (timeout: 30 min) named "amd64_amd64_rootfs"...
 => done, took 0 seconds.
 => Executing "fetch_stage3"
Determining path to latest Gentoo amd64 (amd64) stage3 archive...
 => downloading and processing http://distfiles.gentoo.org/releases/amd64/autobuilds/latest-stage3-amd64.txt
 => Got: 20171208T180422Z/stage3-amd64-20171208T180422Z.tar.bz2
Skipping download of stage3 as it is already present.
verify_gpg(): Warning: $PGP_DIR is empty or not a directory or /root/.gnupg/pubring.gpg
doesen't exist, disabling signature checking!
verify_with_stdutils(): '/var/cache/lxc/gentoo/stage3-amd64-20171208T180422Z/stage3-amd64-20171208T180422Z.tar.bz2': sha512sum OK
verify_with_openssl(): '/var/cache/lxc/gentoo/stage3-amd64-20171208T180422Z/stage3-amd64-20171208T180422Z.tar.bz2': whirlpool OK

Unpacking filesystem from /var/cache/lxc/gentoo/stage3-amd64-20171208T180422Z/stage3-amd64-20171208T180422Z.tar.bz2 to ./test/rootfs ... done.
Will bind mount the portage tree from the host (/usr/portage/tree).
 - writing LXC guest configuration...done.
 - setting hostname...done.
Apply sub_uid/sub_gid
le mode de './usr/lib64/misc/ssh-keysign' a été modifié de 0711 (rwx--x--x) en 4711 (rws--x--x)
...
le mode de './sbin/unix_chkpwd' a été modifié de 0711 (rwx--x--x) en 4711 (rws--x--x)
All done!

You can run your container with the 'lxc-start -f "./test/config" -n "test-nolvm"'
For host-side networking setup info, see "./test/rootfs/etc/conf.d/net"

To enter your container for setup WITHOUT running it, try:
   mount -t proc proc "./test/rootfs/proc"
   mount -o bind /dev "./test/rootfs/dev"
   chroot "./test/rootfs" /bin/bash
   export PS1="(test-nolvm) $PS1"
 (test-nolvm) #   <-- you are now in the guest
... or, start it on the current console with:
   lxc-start -f "./test/config" -n "test-nolvm"
... or, start it elsewhere then use 'lxc-console -n "test-nolvm"' for a shell

(Report bugs to https://github.com/globalcitizen/lxc-gentoo/issues )";

# ls -l test test/rootfs/ 
test:
total 8
-rw-r--r--  1 root     root     3779 14 déc.  23:49 config
drwxr-xr-x 20 10000000 10000000 4096  8 déc.  22:05 rootfs

test/rootfs/:
total 72
drwxr-xr-x  2 10000000 10000000 4096  8 déc.  23:27 bin
drwxr-xr-x  2 10000000 10000000 4096  8 déc.  19:17 boot
drwxr-xr-x  5 10000000 10000000 4096 14 déc.  23:49 dev
drwxr-xr-x 31 10000000 10000000 4096 14 déc.  23:49 etc
drwxr-xr-x  2 10000000 10000000 4096  8 déc.  19:17 home
lrwxrwxrwx  1 10000000 10000000    5  8 déc.  22:04 lib -> lib64
drwxr-xr-x  2 10000000 10000000 4096  8 déc.  22:05 lib32
drwxr-xr-x  9 10000000 10000000 4096  8 déc.  23:27 lib64
drwxr-xr-x  2 10000000 10000000 4096  8 déc.  19:17 media
drwxr-xr-x  2 10000000 10000000 4096  8 déc.  19:17 mnt
drwxr-xr-x  2 10000000 10000000 4096  8 déc.  19:17 opt
drwxr-xr-x  2 10000000 10000000 4096  8 déc.  19:05 proc
drwx------  2 10000000 10000000 4096  8 déc.  19:17 root
drwxr-xr-x  2 10000000 10000000 4096  8 déc.  23:27 run
drwxr-xr-x  2 10000000 10000000 4096  8 déc.  23:29 sbin
drwxr-xr-x  2 10000000 10000000 4096  8 déc.  19:17 sys
drwxrwxrwt  2 10000000 10000000 4096  8 déc.  23:29 tmp
drwxr-xr-x 12 10000000 10000000 4096  8 déc.  23:29 usr
drwxr-xr-x 10 10000000 10000000 4096 14 déc.  23:49 var

# lxc-start -f test/config -n test-nolvm
# lxc-ls -f | grep test-nolvm
test-nolvm       RUNNING 0         -                               192.168.3.99                 -                                      
# lxc-attach -n test-nolvm
test-nolvm / # ls -l 
total 56
drwxr-xr-x   2 root   root   4096  8 déc.  23:27 bin
drwxr-xr-x   2 root   root   4096  8 déc.  19:17 boot
drwxr-xr-x   4 root   root    360 14 déc.  23:55 dev
drwxr-xr-x  31 root   root   4096 14 déc.  23:55 etc
drwxr-xr-x   2 root   root   4096  8 déc.  19:17 home
lrwxrwxrwx   1 root   root      5  8 déc.  22:04 lib -> lib64
drwxr-xr-x   2 root   root   4096  8 déc.  22:05 lib32
drwxr-xr-x   9 root   root   4096  8 déc.  23:27 lib64
drwxr-xr-x   2 root   root   4096  8 déc.  19:17 media
drwxr-xr-x   2 root   root   4096  8 déc.  19:17 mnt
drwxr-xr-x   2 root   root   4096  8 déc.  19:17 opt
dr-xr-xr-x 443 nobody nobody    0 14 déc.  23:55 proc
drwx------   2 root   root   4096  8 déc.  19:17 root
drwxr-xr-x   4 root   root    100 14 déc.  23:55 run
drwxr-xr-x   2 root   root   4096  8 déc.  23:29 sbin
drwxr-xr-x   2 root   root   4096  8 déc.  19:17 sys
drwxrwxrwt   2 root   root     40 14 déc.  23:55 tmp
drwxr-xr-x  12 root   root   4096  8 déc.  23:29 usr
drwxr-xr-x  10 root   root   4096 14 déc.  23:49 var
```

# nolvm(-B) nosub uid/gid(-s) noconfig(-c)
```
# ./lxc-gentoo create -n test-nolvmsubid -u test-nolvmsubid.local -a amd64 -q -i 192.168.3.99/24 -g 192.168.3.254 -P /usr/portage/tree -r ./test/rootfs
NAME           = test-nolvmsubid
UTSNAME        = test-nolvmsubid.local
ROOTFS         = ./test/rootfs
CONFFILE       = /lxc/test-nolvmsubid/config
ARCH           = amd64
ARCHVARIANT    = amd64
STAGE3_TARBALL = 
GUESTROOTPASS  = (not shown)
MIRROR         = http://distfiles.gentoo.org
PORTAGE_SOURCE = /usr/portage/tree
CACHE          = /var/cache/lxc/gentoo
IPV4           = 192.168.3.99/24
IPV4_GATEWAY   = 192.168.3.254
IPV6           = dhcp
IPV6_GATEWAY   = auto
Thanks! Now sit back and relax while your gentoo brews...


Attempting to obtain an exclusive lock (timeout: 30 min) named "amd64_amd64_rootfs"...
 => done, took 0 seconds.
 => Executing "fetch_stage3"
Determining path to latest Gentoo amd64 (amd64) stage3 archive...
 => downloading and processing http://distfiles.gentoo.org/releases/amd64/autobuilds/latest-stage3-amd64.txt
 => Got: 20171208T180422Z/stage3-amd64-20171208T180422Z.tar.bz2
Skipping download of stage3 as it is already present.
verify_gpg(): Warning: $PGP_DIR is empty or not a directory or /root/.gnupg/pubring.gpg
doesen't exist, disabling signature checking!
verify_with_stdutils(): '/var/cache/lxc/gentoo/stage3-amd64-20171208T180422Z/stage3-amd64-20171208T180422Z.tar.bz2': sha512sum OK
verify_with_openssl(): '/var/cache/lxc/gentoo/stage3-amd64-20171208T180422Z/stage3-amd64-20171208T180422Z.tar.bz2': whirlpool OK

Unpacking filesystem from /var/cache/lxc/gentoo/stage3-amd64-20171208T180422Z/stage3-amd64-20171208T180422Z.tar.bz2 to ./test/rootfs ... done.
Will bind mount the portage tree from the host (/usr/portage/tree).
 - writing LXC guest configuration...done.
 - setting hostname...done.
All done!

You can run your container with the 'lxc-start -f "/lxc/test-nolvmsubid/config" -n "test-nolvmsubid"'
For host-side networking setup info, see "./test/rootfs/etc/conf.d/net"

To enter your container for setup WITHOUT running it, try:
   mount -t proc proc "./test/rootfs/proc"
   mount -o bind /dev "./test/rootfs/dev"
   chroot "./test/rootfs" /bin/bash
   export PS1="(test-nolvmsubid) $PS1"
 (test-nolvmsubid) #   <-- you are now in the guest
... or, start it on the current console with:
   lxc-start -f "/lxc/test-nolvmsubid/config" -n "test-nolvmsubid"
... or, start it elsewhere then use 'lxc-console -n "test-nolvmsubid"' for a shell

(Report bugs to https://github.com/globalcitizen/lxc-gentoo/issues )";

# ls -l test test/rootfs/ /lxc/test-nolvmsubid/
/lxc/test-nolvmsubid/:
total 4
-rw-r--r-- 1 root root 3727 14 déc.  23:45 config
-rw-r----- 1 root root    0 14 déc.  23:46 test-nolvmsubid.log

test:
total 4
drwxr-xr-x 20 root root 4096 14 déc.  23:46 rootfs

test/rootfs/:
total 72
drwxr-xr-x  2 root root 4096  8 déc.  23:27 bin
drwxr-xr-x  2 root root 4096  8 déc.  19:17 boot
drwxr-xr-x  5 root root 4096 14 déc.  23:45 dev
drwxr-xr-x 31 root root 4096 14 déc.  23:46 etc
drwxr-xr-x  2 root root 4096  8 déc.  19:17 home
lrwxrwxrwx  1 root root    5  8 déc.  22:04 lib -> lib64
drwxr-xr-x  2 root root 4096  8 déc.  22:05 lib32
drwxr-xr-x  9 root root 4096  8 déc.  23:27 lib64
drwxr-xr-x  2 root root 4096  8 déc.  19:17 media
drwxr-xr-x  2 root root 4096  8 déc.  19:17 mnt
drwxr-xr-x  2 root root 4096  8 déc.  19:17 opt
drwxr-xr-x  2 root root 4096  8 déc.  19:05 proc
drwx------  2 root root 4096  8 déc.  19:17 root
drwxr-xr-x  2 root root 4096  8 déc.  23:27 run
drwxr-xr-x  2 root root 4096  8 déc.  23:29 sbin
drwxr-xr-x  2 root root 4096  8 déc.  19:17 sys
drwxrwxrwt  2 root root 4096  8 déc.  23:29 tmp
drwxr-xr-x 12 root root 4096  8 déc.  23:29 usr
drwxr-xr-x 10 root root 4096 14 déc.  23:45 var

# lxc-start -f /lxc/test-nolvmsubid/config -n test-nolvmsubid
# lxc-ls -f | grep test-nolvm
test-nolvmsubid  RUNNING 0         -                               192.168.3.99                 -                                      
# lxc-attach -n test-nolvmsubid
test-nolvmsubid / # ls -l /
total 56
drwxr-xr-x   2 root root 4096  8 déc.  23:27 bin
drwxr-xr-x   2 root root 4096  8 déc.  19:17 boot
drwxr-xr-x   4 root root  360 14 déc.  23:46 dev
drwxr-xr-x  31 root root 4096 14 déc.  23:46 etc
drwxr-xr-x   2 root root 4096  8 déc.  19:17 home
lrwxrwxrwx   1 root root    5  8 déc.  22:04 lib -> lib64
drwxr-xr-x   2 root root 4096  8 déc.  22:05 lib32
drwxr-xr-x   9 root root 4096  8 déc.  23:27 lib64
drwxr-xr-x   2 root root 4096  8 déc.  19:17 media
drwxr-xr-x   2 root root 4096  8 déc.  19:17 mnt
drwxr-xr-x   2 root root 4096  8 déc.  19:17 opt
dr-xr-xr-x 440 root root    0 14 déc.  23:46 proc
drwx------   2 root root 4096  8 déc.  19:17 root
drwxr-xr-x   4 root root  100 14 déc.  23:46 run
drwxr-xr-x   2 root root 4096  8 déc.  23:29 sbin
drwxr-xr-x   2 root root 4096  8 déc.  19:17 sys
drwxrwxrwt   2 root root   40 14 déc.  23:46 tmp
drwxr-xr-x  12 root root 4096  8 déc.  23:29 usr
drwxr-xr-x  10 root root 4096 14 déc.  23:45 var
test-nolvmsubid / # ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
38: eth0@if39: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    link/ether 4e:3d:7f:b5:c3:b7 brd ff:ff:ff:ff:ff:ff link-netnsid 0
    inet 192.168.3.99/24 brd 192.168.3.255 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fe80::4c3d:7fff:feb5:c3b7/64 scope link 
       valid_lft forever preferred_lft forever
```

J'ai bien fais de faire les tests quelque petit erreur sans -c et une surcharge de valeurr pour ROOTFS
translate: I did the tests, some small error without '-c' and a value overload for ROOTFS

**Comment by specing** [comment]
commit1: there are already configuration file includes provided by the LXC distribution itself, including for Gentoo. Thus it would be good to outsource as much as possible.

commit2: I really, reailly don't like 20-line one-liners, especially when used inside scripts and not the command line. Plus all those 20 lines can be replaced with a single call to uidmapshift, which is a program provided by the LXC people (not sure if it is shipped as part of LXC yet). See https://github.com/specing/lxc-gentoo/blob/master/lxc-gentoo-userns for my user-namespace add-on script.  There are also no checks to see whether valid IDs were supplied by the caller or if the uid changer program terminated with error.

commit3: default values (e.g. for FSSIZE) can be done with one-liners.

**Comment by globalcitizen** [comment]
OK, are we good to merge?

**Comment by slmi** [comment]
It good for me

**Pull request merged**